packages/opentelemetry-scope-async-hooks/src/AsyncHooksScopeManager.ts

```diff
@@ -0,0 +1,158 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ScopeManager } from '@opentelemetry/scope-base';
+import * as asyncHooks from 'async_hooks';
+import { EventEmitter } from 'events';
+import * as shimmer from 'shimmer';
+
+const EVENT_EMITTER_METHODS: Array<keyof EventEmitter> = [
+  'addListener',
+  'on',
+  'once',
+  'prependListener',
+  'prependOnceListener',
+];
+
+export class AsyncHooksScopeManager implements ScopeManager {
+  private asyncHook: asyncHooks.AsyncHook;
+  private scopes: { [uid: number]: unknown } = {};
+  private current: unknown = null;
+
+  constructor() {
+    this.asyncHook = asyncHooks.createHook({
+      init: this.init.bind(this),
+      before: this.before.bind(this),
+      destroy: this.destroy.bind(this),
+    });
+  }
+
+  active(): unknown {
+    return this.current;
+  }
+
+  with<T extends (...args: unknown[]) => ReturnType<T>>(
```

Comment:
There should be a try/catch block to ensure the old scope is restored even if there is an exception. This could be done in base with a private `_with` that would call for example `_enter` and `_exit` on the child class. These methods would handle the specific behavior only for switching scope.

Reply:
I've handled this case but without having `_enter` and `_exit` as i think it's not really needed to have two new concepts in the class. I might be missing a case where it's needed though